### ISSUE TITLE
Deprecation process

### ISSUE THREAD

**Issue by gkurz** [kata-containers/kata-containers#10534]
The kata-containers repo has become fairly large over the years, with new features, new hypervisors, new platforms. Inevitably, even if most features are actively used and maintained by the community, some of them are not, e.g. pmem support for DAX that never gained momentum and is now causing issues (see #10492) or virtio-9p that got replaced by virtio-fs since quite a long time.

Blindly supporting everything is now becoming a real challenge for maintainers. It is about time to consider the removal of some features when they are _obviously_ no longer useful for the project. This cannot happen brutally though as some community members might still be using such features. They should have enough time to engage a discussion around the removal and in the event it is acted, enough time to adapt their workflow.

The QEMU project already has a process in place to handle feature removal. This starts by a discussion to decide if some feature is not wanted anymore. When the decision is taken, this feature is marked as deprecated for next release N in a publicly visible list of _deprecated features_ [1]. The deprecation message should provide suggestions as how to cope with the future removal, e.g. use a newer supported feature. The deprecated feature is guaranteed to still be available for two extra releases. When the development of release N+2 starts, the feature is eligible to be fully removed from the source tree. It is then added to a publicly visible list of _removed features_ [2].

Let's use this issue to discuss how a similar process could be implemented with kata.

Ref:
[1] https://github.com/qemu/qemu/blob/master/docs/about/deprecated.rst
[2] https://github.com/qemu/qemu/blob/master/docs/about/removed-features.rst

**Comment by stevenhorsman** [comment]
Hi Greg, thanks for creating this and I think the clarity of approach from QEMU sounds good with formal deprecated and removal lists. Just a clarification so I can understand the QEMU approach better in terms of timescales:
> The deprecated feature is guaranteed to still be available for two extra releases. When the development of release N+2 starts, the feature is eligible to be fully removed from the source tree.

Does release here refer to major, minor, or fix releases? ie My guess is minor, ie if something is deprecated in 9.0.0 then it would be removed in 9.2.0, which is ~8 months later based on three releases a year I think they do IIRC? I think the key thing to decide is this "notice period" for us as we need to balance removing unmaintained features and stability for the community and downstream.

**Comment by wainersm** [comment]
Hi @gkurz ! I think the description of this issue can even be made into the deprecation process doc!

While I don't think it should be a too bureaucratic process, I think we should consider:

 * How a developer can propose a deprecation request? E.g. open an issue of type "feature deprecation".
   * What information she/he should provide? E.g. obviously the feature and "why", impacts/risks, any known maintainers...etc   
 * How the community is notified about the deprecation intent? E.g. the dev notifies AC and/or messsage `#kata-dev` and/or email kata's mailing list
 * How the deprecation intent is evaluated? E.g. first must pass AC committee then have a grace period for community to manifest
   * How to resolve conflicts? E.g one dev doesn't want the deprecation but at same time the feature cannot be maintained on the expense of the whole community
   * The deprecation is approved. What should be "notice period" (as @stevenhorsman mentioned previously)? When it starts counting?  E.g. begin after the next release after the approval
   * The dev opens a PR. What parts of code she/he should be removing? E.g. code, configuration files, tests, CI jobs...
   * How to document the removal? E.g. Release notes and keep a "removed features" file?

**Comment by sprt** [comment]
To add a bit to what Wainer said, yes I think we should clearly document the communication channels for both:

* The deprecation intent, to give the community time and opportunity to object
* The deprecation approval/rejection, to give the community time and opportunity to adapt

I would say to use at least the mailing list and Slack for both, and most likely GH releases as well. Our community is very much Slack-driven but I believe such announcements are important enough to require surfacing via the mailing list.

**Comment by zvonkok** [comment]
One thing I wanted to add is that components that do not have any CI coverage can be very good candidates for deprecation. Taking the example of 9p, we could remove it today, and the CI would be happy.

**Comment by gkurz** [comment]
> One thing I wanted to add is that components that do not have any CI coverage can be very good candidates for deprecation. Taking the example of 9p, we could remove it today, and the CI would be happy.

On the other hand, 9p can still be useful during development when virtiofs isn't usable, e.g. CoCo. Absence of CI coverage is a hint but not sufficient to decide on deprecation.

**Comment by fidencio** [comment]
Another thing to consider, which was discussed in the AC's meeting on September 25th 2025, is what we do in case one component that we rely on ends up dropping a feature.

In the v3.21.0 release we bumped Cloud Hypervisor, for instance, which removed SGX support. With that we ended up deprecating SGX support when someone is using Cloud Hypervisor as VMM.

Keep in mind, the above is just one example of when it happened, but we should try to have some "recommended" way to go when it happens again.

**Comment by gkurz** [comment]
> Another thing to consider, which was discussed in the AC's meeting on September 25th 2025, is what we do in case one component that we rely on ends up dropping a feature.
> 
> In the v3.21.0 release we bumped Cloud Hypervisor, for instance, which removed SGX support. With that we ended up deprecating SGX support when someone is using Cloud Hypervisor as VMM.
> 
> Keep in mind, the above is just one example of when it happened, but we should try to have some "recommended" way to go when it happens again.

Good point @fidencio. I don't remember the details for CH vs SGX but I see at least two situations :

1) the component has some sort of deprecation process
2) the component abruptly drops a feature without prior notice

(2) means this component isn't to be trusted and we should be very cautious when we claim features that depend on this component.

(1) means that we should carefully monitor release notes of this component and initiate a deprecation process at the kata level. We can also work with the component's community on the deprecation schedule.
